app/models/highlight.rb

```diff
@@ -47,12 +48,118 @@ def normalize_ids
     end
   end
 
+  def page_content_fetchable?
+    openstax_page? && source_metadata && source_metadata.keys.include?('bookVersion')
+  end
+
   protected
 
   def all_mine_from_scope_and_source
     Highlight.where(user_id: user_id).where(scope_id: scope_id).where(source_id: source_id)
   end
 
+  def fetch_page_anchors
+    return unless page_content_fetchable?
+
+    page_content = PageContent.new(
+      book_uuid: scope_id,
+      book_version: source_metadata['bookVersion'],
+      page_uuid: source_id
+    )
+
+    page_content.fetch.anchors
+  end
+
+  def page_anchors
+    @page_anchors ||= fetch_page_anchors
+  end
+
+  def anchor_paths
+    paths = anchor_highlights.pluck(:id, :content_path).select {|h| h[1].present? }
+    paths << [(id || 'new'), content_path] if content_path && !persisted?
+
+    paths.sort_by(&:second)
+  end
+
+  def anchor_highlights
+    all_mine_from_scope_and_source.where(anchor: anchor).order(:order_in_source)
+  end
+
+  def anchor_paths_self_index
+    identifier = id || 'new'
+    anchor_paths.index {|i| i[0] == identifier }
+  end
+
+  def find_prev_content_path_neighbor_id
+    return nil unless anchor_paths_self_index
+    anchor_paths_self_index == 0 ? nil : anchor_paths[anchor_paths_self_index - 1]&.at(0)
+  end
+
+  def find_next_content_path_neighbor_id
+    return nil unless anchor_paths_self_index
+    anchor_paths[anchor_paths_self_index + 1]&.at(0)
+  end
+
+  def has_content_path_neighbor?
+    !!(find_prev_content_path_neighbor_id || find_next_content_path_neighbor_id)
+  end
+
+  def content_paths_comparable?
+    content_path.present? && anchor_paths.many? && has_content_path_neighbor?
+  end
+
+  def page_anchors_comparable?
+    page_anchors.any? && page_anchors.find_index(anchor).present?
+  end
+
+  def find_neighbors
+    return if (prev_highlight_id || next_highlight_id) && neighbors_are_adjacent?
+
+    anchors = all_mine_from_scope_and_source.order(:order_in_source).map(&:anchor)
+    return unless anchors.any?
+
+    # If the highlight is inside an existing anchor, find the neighbors
+    # using content_path, otherwise find it relative to all page anchors
+    if anchors.include?(anchor)
+      set_inside_anchor_neighbors
+    elsif page_anchors_comparable?
+      set_outside_anchor_neighbors
+    elsif !prev_highlight_id && !next_highlight_id
```

Comment:
should this be an `else`? if the prev/next were legitimate we would have short circuited on 116 right?

Reply:
Yes good catch, I think this condition was from before I added the adjacency check up there and is unnecessary.